### PR TITLE
feat(api): Add G-Code to human-readable text parsing

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Support for parsing G-Codes into textual representation

# Changelog

 * Added textual descriptions for all Smoothie G-Codes except for pipette id and model
 * Added parsing into JSON
 * Added parsing program into textual description

# Review requests

None

# Risk assessment

Low to None. This is an external tool that nothing depends on

### CODE REVIEW

File: api/src/opentrons/hardware_control/emulation/smoothie.py [hunk 1]
```diff
@@ -140,8 +151,8 @@ def _handle(self, command: Command) -> Optional[str]:
         func_to_run = self._gcode_to_function_mapping.get(command.gcode)
         return None if func_to_run is None else func_to_run(command)
 
-    @staticmethod
-    def _mount_strings(command: Command) -> Dict[str, str]:
+    @classmethod
+    def _mount_strings(cls, command: Command) -> Dict[str, str]:
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/emulation/smoothie.py]: Why this change?

File: api/src/opentrons/hardware_control/emulation/smoothie.py [hunk 2]
```diff
@@ -155,8 +166,14 @@ def _mount_strings(command: Command) -> Dict[str, str]:
         Returns:
             A dict of L and/or R to the string value following it.
         """
-        pars = (i.groupdict() for i in
-                SmoothieEmulator.WRITE_INSTRUMENT_RE.finditer(command.body))
-        result = {p['mount']: p['value'] for p in pars}
+        pars = [i.groupdict() for i in
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/emulation/smoothie.py]: Why change to a list comprehension?

File: api/src/opentrons/hardware_control/emulation/smoothie.py [hunk 3]
```diff
@@ -155,8 +166,14 @@ def _mount_strings(command: Command) -> Dict[str, str]:
...
+                SmoothieEmulator.WRITE_INSTRUMENT_RE.finditer(command.body)]
+        result = {
+            p['mount']: p['value'] + '0' * (
+                cls.INSTRUMENT_AND_MODEL_STRING_LENGTH - len(p['value'])
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/emulation/smoothie.py]: Why do we have to pad with zeros?

File: api/src/opentrons/hardware_control/emulation/smoothie.py [hunk 4]
```diff
@@ -23,6 +23,7 @@ class SmoothieEmulator(AbstractEmulator):
     """Smoothie emulator"""
 
     WRITE_INSTRUMENT_RE = re.compile(r"(?P<mount>[LR])\s*(?P<value>[a-f0-9]+)")
+    INSTRUMENT_AND_MODEL_STRING_LENGTH = 64
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/emulation/smoothie.py:26]: Thank you!

File: api/src/opentrons/hardware_control/g_code_parsing/g_code.py [hunk 5]
```diff
@@ -1,13 +1,77 @@
...
 
 
 class GCode:
     """
     Middleware class to provide functionality to define G-Codes as well as
     convert them to human-readable JSON form
     """
+
+    G_CODE_EXPLANATION_MAPPING = {
+        # Weird Enum thing stopping me from using values from
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code.py:22]: Not for this PR. But now that all the drivers use enums, perhaps we can think about making these lookups a part of a base enum type for GCODES. 

For the future.....

**Review by amitlissack — COMMENTED** [review]
Part of the way through...

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 1 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py:155]: Sorry this was a change when I was debugging that I forgot to change back.
Will address now.

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 2 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py]: Sorry this was a change when I was debugging that I forgot to change back.
Will address now.

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 3 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py:173]: Because that is how it is normally stored. If you look at [string_to_hex](https://github.com/Opentrons/opentrons/blob/edge/api/src/opentrons/drivers/utils.py#L151) which is called in the constructor of the emulator, all values are padded to 64 characters in length

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 1 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py:155]: Sorry this was a change when I was debugging that I forgot to change back.
Will address now.

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 2 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py:169]: Sorry this was a change when I was debugging that I forgot to change back.
Will address now.

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 3 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py]: Because that is how it is normally stored. If you look at string_to_hex which is called in the constructor of the emulator, all values are padded to 64 characters in length

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 1 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py:155]: Sorry this was a change when I was debugging that I forgot to change back.
Will address now.

Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/emulation/smoothie.py]: Oh this was needed to be able to get the INSTURMENT_AND_MODEL_STRING_LENGTH constant

Comment (amitlissack) [reply api/src/opentrons/hardware_control/emulation/smoothie.py]: Ah. An alternative would be to use the `<classname>.INSTRUMENT_AND_MODEL_STRING_LENGTH`.

File: api/src/opentrons/hardware_control/emulation/smoothie.py (hunk 3 above)
Comment (amitlissack) [reply api/src/opentrons/hardware_control/emulation/smoothie.py]: I see. I never thought that was necessary, but it makes sense when you're trying to validate that gcodes match.

File: api/src/opentrons/hardware_control/g_code_parsing/g_code.py [hunk 6]
```diff
@@ -1,13 +1,86 @@
...
+import re
+from typing import List
+
 from .errors import UnparsableGCodeError
 from opentrons.drivers.smoothie_drivers.driver_3_0 import GCODE as SMOOTHIE_GCODE
 from opentrons.hardware_control.g_code_parsing.utils import reverse_enum
+from opentrons.hardware_control.emulation.parser import Parser
+from opentrons.hardware_control.g_code_parsing.g_code_functionality_defs.\
+    g_code_functionality_def_base import Explanation
+from .g_code_functionality_defs.smoothie import (
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code.py]: Not asking for a change. Just pointing out that you could do `from .g_code_functionality_defs import smoothie` to avoid a crazy long import line. Then refer to these types as `smoothie.MoveGCodeFunctionalityDef` for example.

Personal taste really.

File: api/src/opentrons/hardware_control/g_code_parsing/g_code.py [hunk 7]
```diff
@@ -21,22 +94,45 @@ class GCode:
...
+    ]
+
+    @classmethod
+    def from_raw_code(cls, raw_code: str, device: str, response: str) -> List[GCode]:
+        g_code_list = []
+        for g_code in Parser().parse(raw_code):
+            if g_code.gcode not in cls.SPECIAL_HANDLING_REQUIRED_G_CODES:
+                g_code_list.append(cls(device, g_code.gcode, g_code.params, response))
+            else:
+                left_or_right = g_code.body.strip()[0]
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code.py:125]: It may not be clear to people why this is here. Could you add a little comment about this special case?

File: api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/g_code_functionality_def_base.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    COMMAND_EXPLANATION_KEY = 'Command Explanation'
+    RESPONSE_KEY = 'Response'
+
+    code: str
+    command_name: str
+    provided_args: Dict[str, Any]
+    command_explanation: str
+    response: str
+
+    def to_dict(self) -> Any:
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/g_code_functionality_def_base.py]: The return type could be more specific. Not criticial.

File: api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/acceleration_g_code_functionality_def.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+from typing import Dict
+from string import Template
+from enum import Enum
+from opentrons.hardware_control.g_code_parsing.g_code_functionality_defs.\
+    g_code_functionality_def_base import GCodeFunctionalityDefBase
+
+
+class AccelerationGCodeFunctionalityDef(GCodeFunctionalityDefBase):
+    EXPECTED_ARGS = ['S', 'X', 'Y', 'Z', 'A', 'B', 'C']
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/acceleration_g_code_functionality_def.py:10]: Since `ValDefinedMessage` seems to define a value per expected arg does `EXPECTED_ARGS` need to exist? Can we instead use the names in the enum instead?

File: api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/move_g_code_functionality_def.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+from typing import Dict
+from string import Template
+from enum import Enum
+from opentrons.hardware_control.g_code_parsing.g_code_functionality_defs.\
+    g_code_functionality_def_base import GCodeFunctionalityDefBase
+
+
+class MoveGCodeFunctionalityDef(GCodeFunctionalityDefBase):
+    EXPECTED_ARGS = ['X', 'Y', 'Z', 'A', 'B', 'C', 'F']
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/move_g_code_functionality_def.py:10]: Same comment about possibly redundant list as above.

**Review by amitlissack — COMMENTED** [review]
This is really solid looking. Great job. 

I left a lot of comments which are mostly style / personal taste variety. 

But please leave comments explaining some special exceptional cases (as in the `left_or_right` bit).

File: api/src/opentrons/hardware_control/g_code_parsing/g_code.py (hunk 6 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/g_code_parsing/g_code.py]: That made it look much cleaner. Thank you!

File: api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/acceleration_g_code_functionality_def.py (hunk 9 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/g_code_parsing/g_code_functionality_defs/smoothie/acceleration_g_code_functionality_def.py:10]: I am using that list to maintain a special ordering. I will add a comment explaining that

File: api/src/opentrons/hardware_control/g_code_parsing/g_code.py (hunk 7 above)
Comment (amitlissack) [reply api/src/opentrons/hardware_control/g_code_parsing/g_code.py:125]: Thank you

**Review by amitlissack — APPROVED** [review]
🥇
